pyomo/gdp/plugins/partition_disjuncts.py

```diff
@@ -0,0 +1,850 @@
+from pyomo.network import Port
+from pyomo.common.collections import ComponentSet
+from pyomo.repn import generate_standard_repn
+from pyomo.core.expr import current as EXPR
+from pyomo.opt import SolverFactory
+from pyomo.util.vars_from_expressions import get_vars_from_components
+
+from pyomo.gdp import Disjunct, Disjunction, GDP_Error
+from pyomo.gdp.util import (is_child_of, _to_dict, verify_successful_solve,
+                            NORMAL, clone_without_expression_components,
+                            _warn_for_active_disjunct, get_gdp_tree)
+from pyomo.core.util import target_list
+from pyomo.contrib.fbbt.fbbt import compute_bounds_on_expr
+from weakref import ref as weakref_ref
+
+from math import floor
+
+import logging
+logger = logging.getLogger('pyomo.gdp.partition_disjuncts')
+
+NAME_BUFFER = {}
+
+def _generate_additively_separable_repn(nonlinear_part):
+    if nonlinear_part.__class__ is not EXPR.SumExpression:
+        # This isn't separable, so we just have the one expression
+        return {'nonlinear_vars': [tuple(v for v in EXPR.identify_variables(
+            nonlinear_part))], 'nonlinear_exprs': [nonlinear_part]}
+
+    # else, it was a SumExpression, and we will break it into the summands,
+    # recording which variables are there.
+    nonlinear_decomp = {'nonlinear_vars': [],
+                        'nonlinear_exprs': []}
+    for summand in nonlinear_part.args:
+        nonlinear_decomp['nonlinear_exprs'].append(summand)
+        nonlinear_decomp['nonlinear_vars'].append(
+            tuple(v for v in EXPR.identify_variables(summand)))
+
+    return nonlinear_decomp
+
+def arbitrary_partition(disjunction, P):
+    """
+    Returns a valid partition into P sets of the variables that appear in
+    algebraic additively separable constraints in the Disjuncts in
+    'disjunction'. Note that this method may return an invalid partition
+    if the constraints are not additively separable!
+
+    Arguments:
+    ----------
+    disjunction : A Disjunction object for which the variable partition will be
+                 created.
+    P : An int, the number of partitions
+    """
+    # collect variables
+    v_set = ComponentSet()
+    for disj in disjunction.disjuncts:
+        v_set.update(get_vars_from_components(disj, Constraint,
+                                              descend_into=Block, active=True))
+    # assign them to partitions
+    partitions = []
+    V = len(v_set)
+    whole = floor(V/P)
+    for partition in range(V % P):
+        partitions.append(ComponentSet())
+        # add whole + 1 vars
+        for i in range(whole + 1):
+            partitions[partition].add(v_set.pop())
+    # for the rest, add whole vars
+    for partition in range(V % P, P):
+        partitions.append(ComponentSet())
+        for i in range(whole):
+            partitions[partition].add(v_set.pop())
+
+    return partitions
+
+def compute_optimal_bounds(expr, global_constraints, opt):
```

Comment:
Yes! I'd like to discuss this before we merge because at the very least we should make sure that this method is designed so we can extend to use coramin.